e2e: Moved to Cortex e2e test and GithubActions.

bwplotka opened the pull request:
This PR moves all Thanos e2e tests to the new [Cortex e2e framework](https://github.com/cortexproject/cortex/tree/master/integration/e2e) written by amazing @pracucci . 

Our e2e tests are responsible for spinning Thanos, Prometheus, Alertmanager, and Minio components by executing pre-built binary and invoking some more or less complex tests. 

### Why changing?

Previously our e2e tests were using our written scheduler which schedules all components as children processes. It was also running on CircleCI inside docker(!). This was problematic for various reasons:

* It was harder to clean up on crash.
* Process isolation was weaker: We couldn't isolate resources (e.g network: everything could talk to everything). This also mean measuring performance was harder. 
* It was highly unreliable/flaky, mostly due to CircleCI free tier being ultra slow.

Fixes: https://github.com/thanos-io/thanos/issues/2046
Fixes: https://github.com/thanos-io/thanos/issues/2045
Partially: https://github.com/thanos-io/thanos/issues/1704

### Make e2e great again!

@pracucci recently started [this e2e library](https://github.com/cortexproject/cortex/tree/master/integration/e2e) for Cortex need. Simplicity and the fact that it is docker based, motivated me to extend it a bit and use it for Thanos. 

* Now it is **much faster**  and should be not flaky.
* It is much cleaner (IMO). Less LOCs definitely.
* It is based on metrics so we tests our instrumentation on the way! We test not only correctness but also usefulness! You will like it @kakkoyun ;p
* It is scheduling everything in docker, so we can isolate resources better. E.g limits, ports. Also, we can monitor those better. :muscle: 

### Changes:

* Excluded it from CircleCI and moved to GithubAction: *mucho más rápido*.
* Rewrote e2e tests for Cortex e2e library. Now all is running in docker containers.
* Tweaked our tests to be based on **metrics** instead of querying. Also Tweaked flags for faster intervals, discovery, etc.
* Fixed logs for certain components.
* Refactor discovery a bit. Functionally should be the same.

Kudos for @pracucci for starting and making it so clean :muscle: and possible for Thanos to reuse! :tada: 

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

bwplotka commented:
This should be good to go!

Sorry for big PR but hopefully it should be faster, more reliable and cleaner.

### Some extra learnings from this work:

* Receiver is like extremely spammy and DoS-ed with default backoff retry of Prometheus remote write (100ms-300ms).
This is way too extreme. Either we need to increase client backoff or return
status code that will stop retry for certain errors (e.g misconfiguration). Or we should delegate this to the rate limit service.
* When a block is removed we still fill the cache with old entries from it. Not functionally wrong, just silly.
* Receiver with outage makes other recivers do this:
```
receive-1: level=error name=receive-1 ts=2020-02-18T17:30:35.341203944Z caller=handler.go:411 component=receive component=receive-handler msg="forwarding request" err="rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = \"transport: Error while dialing dial tcp: lookup e2e_test_receive_replication_with_outage-receive-3 on 127.0.0.11:53: no such host\"" endpoint=e2e_test_receive_replication_with_outage-receive-3:9091
receive-1: level=error name=receive-1 ts=2020-02-18T17:30:35.341446399Z caller=handler.go:411 component=receive component=receive-handler msg="forwarding request" err="rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = \"transport: Error while dialing dial tcp: lookup e2e_test_receive_replication_with_outage-receive-3 on 127.0.0.11:53: no such host\"" endpoint=e2e_test_receive_replication_with_outage-receive-3:9091
receive-1: level=error name=receive-1 ts=2020-02-18T17:30:35.341405191Z caller=handler.go:411 component=receive component=receive-handler msg="forwarding request" err="rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = \"transport: Error while dialing dial tcp: lookup e2e_test_receive_replication_with_outage-receive-3 on 127.0.0.11:53: no such host\"" endpoint=e2e_test_receive_replication_with_outage-receive-3:9091
receive-1: level=error name=receive-1 ts=2020-02-18T17:30:35.348055804Z caller=handler.go:411 component=receive component=receive-handler msg="forwarding request" err="rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = \"transport: Error while dialing dial tcp: lookup e2e_test_receive_replication_with_outage-receive-3 on 127.0.0.11:53: no such host\"" endpoint=e2e_test_receive_replication_with_outage-receive-3:9091
receive-1: level=error name=receive-1 ts=2020-02-18T17:30:35.34820348Z caller=handler.go:411 component=receive component=receive-handler msg="forwarding request" err="rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = \"transport: Error while dialing dial tcp: lookup e2e_test_receive_replication_with_outage-receive-3 on 127.0.0.11:53: no such host\"" endpoint=e2e_test_receive_replication_with_outage-receive-3:9091
receive-1: level=error name=receive-1 ts=2020-02-18T17:30:35.349051032Z caller=handler.go:411 component=receive component=receive-handler msg="forwarding request" err="rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = \"transport: Error while dialing dial tcp: lookup e2e_test_receive_replication_with_outage-receive-3 on 127.0.0.11:53: no such host\"" endpoint=e2e_test_receive_replication_with_outage-receive-3:9091
receive-1: level=error name=receive-1 ts=2020-02-18T17:30:36.314702005Z caller=handler.go:411 component=receive component=receive-handler msg="forwarding request" err="rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = \"transport: Error while dialing dial tcp: lookup e2e_test_receive_replication_with_outage-receive-3 on 127.0.0.11:53: no such host\"" endpoint=e2e_test_receive_replication_with_outage-receive-3:9091
receive-1: level=error name=receive-1 ts=2020-02-18T17:30:36.314772679Z caller=handler.go:411 component=receive component=receive-handler msg="forwarding request" err="rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = \"transport: Error while dialing dial tcp: lookup e2e_test_receive_replication_with_outage-receive-3 on 127.0.0.11:53: no such host\"" endpoint=e2e_test_receive_replication_with_outage-receive-3:9091
receive-1: level=error name=receive-1 ts=2020-02-18T17:30:36.315003796Z caller=handler.go:411 component=receive component=receive-handler msg="forwarding request" err="rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = \"transport: Error while dialing dial tcp: lookup e2e_test_receive_replication_with_outage-receive-3 on 127.0.0.11:53: no such host\"" endpoint=e2e_test_receive_replication_with_outage-receive-3:9091
receive-1: level=error name
``` x 1000
Is such retry without back-off really makes sense? @brancz @squat @metalmatze

bwplotka commented:
cc @pracucci

bwplotka commented:
Kudos for @pracucci for starting this e2e library and making it so clean :muscle:

bwplotka commented:
**BTW Our e2e test now takes 4m instead of ~15m!!! :heart: **

bwplotka commented:
Also @simonpasquier could you look on `rule_test.go` part?

metalmatze commented on pkg/discovery/dns/provider.go:
```diff
@@ -23,7 +24,7 @@ type Provider struct {
 	logger   log.Logger
 
-	resolverAddrs         *prometheus.GaugeVec
+	resolverAddrs         *extprom.TxGaugeVec
```
Is this PR building on some PR that has not yet been merged? Or why is this change in here, when "just" updating the CI? :thinking:

bwplotka replied:
What do you mean? We need that for metric to be "atomic"

brancz commented on cmd/thanos/receive.go:
```diff
@@ -318,6 +321,7 @@ func runReceive(
 		if cw != nil {
 			ctx, cancel := context.WithCancel(context.Background())
 			g.Add(func() error {
+				// TODO: If config is empty we never start receiver. We might want to fail in this case.
```
can we open an issue about this? :)

bwplotka replied:
sure (:

bwplotka replied:
Done: https://github.com/thanos-io/thanos/issues/2160

bwplotka commented:
Is this good to go? (: cc @squat

kakkoyun commented on test/e2e/rule_test.go:
```diff
@@ -197,32 +176,39 @@ func (m *mockAlertmanager) ServeHTTP(resp http.ResponseWriter, req *http.Request
-func TestRuleAlertmanagerHTTPClient(t *testing.T) {
-	a := newLocalAddresser()
+func TestRule_AlertmanagerHTTPClient(t *testing.T) {
+	t.Skip("TODO: Allow HTTP ports from binaries running on host to be accessible.")
```
Is this `skip` intentional or forgotten? If it's intentional better to create an issue, right?
Otherwise, we're skipping a test that previously functional.

bwplotka replied:
We are indeed.

kakkoyun commented on test/e2e/rule_test.go:
```diff
@@ -296,577 +264,226 @@ func TestRuleAlertmanagerHTTPClient(t *testing.T) {
-			// We do expect warnings.
-			return errors.Errorf("unexpected number of warnings, expected 1, got %s", warnings)
-		}
+	t.Skip("TODO: Allow HTTP ports from binaries running on host to be accessible.")
```
again we should have an issue for this as well.

kakkoyun commented on pkg/discovery/dns/provider_test.go:
```diff
@@ -37,40 +36,67 @@ func TestProvider(t *testing.T) {
 	result = prv.Addresses()
 	sort.Strings(result)
 	testutil.Equals(t, ips, result)
+	testutil.Equals(t, 3, promtestutil.CollectAndCount(prv.resolverAddrs))
```
Could we please add a comment in this test case to clarify the repetition of test cases. I believe, essentially it is for testing determinist resolution.

bwplotka replied:
Wait. I cannot see the repetition.. We check same stuff and we cache a lot, so it's definitely not repetition.

bwplotka commented:
Thanks for the review @kakkoyun!

GiedriusS commented on Makefile:
```diff
@@ -278,19 +276,21 @@ web-pre-process:
+.PHONY: lint
+lint: ## Runs various static analysis against our code.
+lint: check-git deps $(GOLANGCILINT) $(MISSPELL)
+	$(call require_clean_work_tree,"detected not clean muster before running lint")
```
```suggestion
	$(call require_clean_work_tree,"detected not clean master before running lint")
```

GiedriusS commented on test/e2e/e2ethanos/services.go:
```diff
@@ -0,0 +1,315 @@
+	return prom, sidecar, nil
+}
+
+func generateFileSD(addresses []string) string {
```
I guess this is easier than constructing the whole thing with a proper type :smile:

bwplotka replied:
:+1:

GiedriusS commented on test/e2e/query_test.go:
```diff
@@ -56,127 +63,109 @@ func sortResults(res model.Vector) {
+	t.Helper()
 
-	// With deduplication.
+	fmt.Println("queryAndAssert: Waiting for", len(expected), "results for query", query)
```
`fmt`? :eye: Ah, but I see that the cortex framework is using it as well :/

GiedriusS commented on test/e2e/store_gateway_test.go:
```diff
@@ -5,182 +5,242 @@ package e2e_test
 )
 
+// TODO(bwplotka): Extend this test to have multiple stores and memcached.
+// TODO(bwplotka): Extent this test for downsampling.
```
```suggestion
// TODO(bwplotka): Extend this test for downsampling.
```

GiedriusS commented on Makefile:
```diff
@@ -133,126 +136,122 @@ assets: $(GOBINDATA)
-# It installs supported versions of Prometheus and alertmanager to test against in e2e.
+.PHONY: test-e2e
+test-e2e: ## Runs all Thanos e2e docker-based e2e tests from test/e2e.
+test-e2e: docker
```
Can we add a message here before running everything that you either need to have `root` or set up your user in a certain way?

bwplotka replied:
done.

GiedriusS commented on test/e2e/receive_test.go:
```diff
@@ -8,222 +8,227 @@ import (
-		// succeed and the time series should be replicated to the other nodes.
-		receiveGRPC1, receiveGRPC2, receiveGRPC3 := a.New(), a.New(), a.New()
+		// Recreate again, but with hashring config.
+		r1, err = e2ethanos.NewReceiver(s.SharedDir(), s.NetworkName(), "1", 3, h)
```
Shouldn't we close the previous ones before trying to start these up? AFAICT it does an equivalent of `docker run ...` later on and it doesn't stop the previously created ones.

bwplotka replied:
`s.Close` does it.

GiedriusS commented:
Circle CI also prints:
```
error while removing sharedDir /tmp/e2e_integration_test569732206 err: unlinkat /tmp/e2e_integration_test569732206/data/rule/1/wal/00000000: permission denied
```
Worth investigating? Haven't checked myself but it might point to some potential problems.

bwplotka commented on go.mod:
```diff
@@ -115,6 +65,7 @@ require (
 replace (
 	// Mitigation for: https://github.com/Azure/go-autorest/issues/414
 	github.com/Azure/go-autorest => github.com/Azure/go-autorest v12.3.0+incompatible
+	github.com/cortexproject/cortex => github.com/bwplotka/cortex v0.0.0-20200218165228-c04fa1c09090
```
:+1:

bwplotka commented on pkg/store/bucket.go:
```diff
@@ -368,7 +368,7 @@ func (s *BucketStore) SyncBlocks(ctx context.Context) error {
 			s.metrics.blockDropFailures.Inc()
 		}
-		level.Debug(s.logger).Log("msg", "dropped outdated block", "block", id)
+		level.Info(s.logger).Log("msg", "unloaded outdated block", "block", id)
```
Fair (:

bwplotka commented on test/e2e/e2ethanos/services.go:
```diff
@@ -0,0 +1,315 @@
+		promImage,
+		e2e.NewCommandWithoutEntrypoint("prometheus", e2e.BuildArgs(map[string]string{
+			"--config.file":                     filepath.Join(container, "prometheus.yml"),
+			"--storage.tsdb.path":               filepath.Join(container),
```
xD

bwplotka replied:
Nicely spotted :heart:

bwplotka commented:
Addressed all comments and got positive feedback. @kakkoyun approved, so let's merge it to stop rerunning CI all the time because of flakinesses... @yeya24 @GiedriusS @metalmatze , hope that's ok (: :hugs:
